### ISSUE TITLE
Support Swap

### ISSUE THREAD

**Issue by joelchen** [bottlerocket-os/bottlerocket#1911]
**What I'd like:**
Support enabling swap via Linux commands like fallocate, mkswap, and swapon. Using swap allows the container to write excess memory requirements to disk when the container has exhausted all the RAM that is available to it.

**Any alternatives you've considered:**
Have submitted an issue on aws-cdk to enable swap through aws-cdk: https://github.com/aws/aws-cdk/issues/18460

**Comment by mjgp2** [comment]
I am also interested in this.

I am wondering if it is possible to have a script using the ephemeral disks approach to attach an EBS volume and format & mount it as swap?

https://github.com/bottlerocket-os/bottlerocket/pull/1601

**Comment by zmrow** [comment]
@joelchen @mjgp2 Out of curiosity, could either of you share some additional details around your setup and your need for swap?

My first thought is that a script in a [bootstrap container](https://github.com/bottlerocket-os/bottlerocket/discussions/1618#discussioncomment-880690) could help with this, but I haven't personally tried it yet.  I don't know off the top of my head if bootstrap containers have the permissions to do so.  I'll try it out this week, or if either of you beat me to it let me know!

**Comment by joelchen** [comment]
@zmrow My setup is pretty simple using aws-cdk, added LinuxParams from @mjgp2's advice in https://github.com/aws/aws-cdk/issues/18460, so that my Rust application does not crash with 137 error when it consumes more than 1 GiB of memory available in t4g.micro instance:
```
class LinuxParams extends ecs.LinuxParameters {
  renderLinuxParameters(): ecs.CfnTaskDefinition.LinuxParametersProperty {
    const p = super.renderLinuxParameters() as any;
    p.maxSwap = 1024;
    p.swappiness = 60;
    return p;
  }
}

const vpc = new ec2.Vpc(this, 'Vpc', { maxAzs: 1, natGateways: 0, subnetConfiguration: [
  {
    name: 'Public',
    subnetType: ec2.SubnetType.PUBLIC
  },
] });

const cluster = new ecs.Cluster(this, 'EcsCluster', { vpc });

cluster.addCapacity('AutoScalingGroup', {
  maxCapacity: 1,
  instanceType: new ec2.InstanceType('t4g.micro'),
  machineImageType: ecs.MachineImageType.BOTTLEROCKET
});

const taskDefinition = new ecs.Ec2TaskDefinition(this, 'TaskDef');

taskDefinition.addContainer('Container', {
  image: ecs.ContainerImage.fromAsset('./application'),
  memoryReservationMiB: 512,
  linuxParameters: new LinuxParams(this, 'L')
});

const ecsService = new ecs.Ec2Service(this, 'Service', {
  cluster,
  taskDefinition,
});
```

**Comment by zmrow** [comment]
Thanks for the additional details!

I had some time yesterday to test my theory using bootstrap containers and found it's not possible.  The reason it's not possible is that we use the [default seccomp profile](https://docs.docker.com/engine/security/seccomp/) when [setting the container options](https://github.com/bottlerocket-os/bottlerocket/blob/f1a216200947f67470db9133102ac7ea7108bbb5/sources/host-ctr/cmd/host-ctr/main.go#L642) for the bootstrap container.  The default seccomp profile explicitly blocks the `swapon/off` syscalls.

We'll keep this ask in mind if we find other reasons to relax the security policy.

**Comment by chlunde** [comment]
Another use case is for nodes running builds (such as tekton and github actions runners).

Typically only a few builds are memory hungry, but it's difficult for the end user to provision the exact memory required, and it would be expensive to provision memory for the largest build.

In this case it's better to have swap and overprovisioning of memory. Then we get slow builds when two memory hungry builds are running concurrently, instead of failed builds.

**Comment by stefansundin** [comment]
In my opinion this issue should probably stay open until there's a configuration option to set up swap (e.g. a list of `[[swap]]` configurations that let you configure multiple swap files each with a location and size). And the bootstrap container approach currently only works with ECS and not Kubernetes.

**Comment by easy1481437320** [comment]
i also intereste enable the swap at bottlerocket-os for  EKS nodegroup,  i try to follow https://github.com/bottlerocket-os/bottlerocket/pull/2537  add below config at AWS EC2 lanuch template  user-data and create the EKS nodegroup but failed.

```
[settings.bootstrap-containers.swap]
mode = "always"
essential = false
user-data = "IyEvYmluL2Jhc2ggLWV4CmlmIFtbICEgLWYgLy5ib3R0bGVyb2NrZXQvcm9vdGZzL3Zhci9zd2FwZmlsZSBdXTsgdGhlbgogIGRkIGlmPS9kZXYvemVybyBvZj0vLmJvdHRsZXJvY2tldC9yb290ZnMvdmFyL3N3YXBmaWxlIGJzPTFNIGNvdW50PTE2Mzg0CiAgY2htb2QgNjAwIC8uYm90dGxlcm9ja2V0L3Jvb3Rmcy92YXIvc3dhcGZpbGUKICBta3N3YXAgLy5ib3R0bGVyb2NrZXQvcm9vdGZzL3Zhci9zd2FwZmlsZQpmaQpzd2Fwb24gLy5ib3R0bGVyb2NrZXQvcm9vdGZzL3Zhci9zd2FwZmlsZQ=="
```
```

**Comment by stefansundin** [comment]
@easy1481437320 I don't think this works with EKS. In my comment above I wrote `And the bootstrap container approach currently only works with ECS and not Kubernetes.` and I assume that is still the case.

And you forgot to include `source = "public.ecr.aws/stefansundin/bottlerocket-bootstrap-exec-user-data:latest"`, and also the error message or logs or anything really that would help debug the problem. But the problem is probably that it won't work with Kubernetes.

**Comment by awoimbee** [comment]
@stefansundin just to be clear, the blocker is that bottlerocket needs to pass `--fail-swap-on=false` to kubelet, either:

* Via arbitrary flags: https://github.com/bottlerocket-os/bottlerocket/issues/1140
* Via a new config key
 
Or we could also wait for another 2~3 years for kubelet to be able to start without needing `--fail-swap-on=false` (depends on https://github.com/kubernetes/enhancements/issues/2400)
EDIT: looks like k8s 1.30 will change stuff about swap

Am I right ?

**Comment by stefansundin** [comment]
@awoimbee I think you are correct. I am probably the wrong person to ask though as I have never used bottlerocket with kubernetes.

**Comment by z0rc** [comment]
> he blocker is that bottlerocket needs to pass `--fail-swap-on=false` to kubelet

Fixed with https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/473.

**Comment by z0rc** [comment]
@ytsssun please reopen. Linked issue isn't a complete implementation of swap support. https://github.com/bottlerocket-os/bottlerocket/issues/1911#issuecomment-1302800686 still stands.
